core/src/layers/observe/metrics.rs

```diff
@@ -132,18 +157,60 @@ impl<A: Access, I: MetricsIntercept> Layer<A> for MetricsLayer<I> {
     fn layer(&self, inner: A) -> Self::LayeredAccess {
         let info = inner.info();
 
+        // Update http client with metrics http fetcher.
+        info.update_http_client(|client| {
+            HttpClient::with(MetricsHttpFetcher {
+                inner: client.into_inner(),
+                info: info.clone(),
+                interceptor: self.interceptor.clone(),
+            })
+        });
+
         MetricsAccessor {
-            inner: Arc::new(inner),
+            inner,
             info,
             interceptor: self.interceptor.clone(),
         }
     }
 }
 
+/// The metrics http fetcher for opendal.
+pub struct MetricsHttpFetcher<I: MetricsIntercept> {
+    inner: HttpFetcher,
+    info: Arc<AccessorInfo>,
+    interceptor: I,
+}
+
+impl<I: MetricsIntercept> HttpFetch for MetricsHttpFetcher<I> {
+    async fn fetch(&self, req: http::Request<Buffer>) -> Result<http::Response<HttpBody>> {
+        // Extract context from the http request.
+        let size = req.body().len();
+        let op = req.extensions().get::<Operation>().copied();
+
+        let start = Instant::now();
+        let res = self.inner.fetch(req).await;
+        let duration = start.elapsed();
+
+        // We only inject the metrics when the operation exists.
+        if let Some(op) = op {
+            if res.is_ok() {
+                self.interceptor.observe_http_request_duration_seconds(
+                    self.info.clone(),
+                    op,
+                    duration,
+                );
+                self.interceptor
+                    .observe_http_request_bytes(self.info.clone(), op, size);
+            }
```

Comment:
Is it necessary to add `http_request_errors_total` counter metric here when `res.is_err()`?

Reply:
Should already been covered by operation errors.